### DISCUSSION TITLE
WearOS 5 compatability discussion (Galaxy Watch 7 & Pixel Watch 3)

### DISCUSSION THREAD

**Discussion by gth001** [NightscoutFoundation/xDrip#3581] [Bug reports]
Update: https://github.com/NightscoutFoundation/xDrip/discussions/3581#discussioncomment-13015219

OLD THREAD BELOW:
——

I bought a Samsung Watch 7 (only $59 on Prime day!), but sadly the xDrip watchface gives a "made for an older OS" error. The xDrip prefs, snooze, etc all work on the watch, just no watchfaces. It appears Google announced a year ago that all older watchfaces not following the new "Watch Face Format (WFF)" cannot be used on any watch that ships with Wear OS 5.

Are there any short term plans to add compatibility? Or am I better off not holding my breath, and just return this watch to go back to my old one?

- Other xDrip users who mistakenly bought a WearOs 5 watch https://www.facebook.com/groups/xDripG5/posts/26014529674860515/
- News report https://9to5google.com/2024/07/10/wear-os-5-watch-face-format/
- Developer documentation https://developer.android.com/training/wearables/wff
- https://android-developers.googleblog.com/2023/05/introducing-watch-face-format-for-wear-os.html

Either way, thanks for all the hard work everyone does!

**Comment by gth001** [comment]
For now I plan to return the watch for a refund in week or two. I don't have much hope of getting a graph watchface working.

**Reply by KeijoMika** [reply]
This change in new Wear OS really sucks for us. But there is some hope. I got my very heavily modified graph working by modifying the source code and saving the graph into Bitmap format when it updates. Then I updated the Customcomplicationproviderservice to provide this image. Next, I made a watchface that uses SMALL_IMAGE complication using the WatchFaceStudio. 

Maybe some developer might get ideas from how I approached it, or comes up with even a better, more efficient way if possible. I don't feel comfortable sharing my changes as they have not been tested thoroughly and heavily modify the xDrip source code. People's lives might depend on it.

**Reply by gth001** [reply]
That is good to hear. At least it isn't impossible then. 

I don't have the skills to modify xDrip, but perhaps I will tinker with Watch Face Studio a bit, before refunding this watch. It looks like I should be able to display mg/dL, delta, and trend arrow at the least. Perhaps I can make this work for myself. I suppose the graph isn't vital for me.

![image](https://github.com/user-attachments/assets/282bffc5-5d35-4f89-947c-4fcd231a48b9)

**Comment by gth001** [comment]
I can't figure out how to increase the font size of the complication.

It's always a microscopically tiny font, no matter how big I make the font of everything in Watchface Studio.

It's like xdrip has the font size hard coded, or Watchface Studio has a bug

<img width="368" alt="image" src="https://github.com/user-attachments/assets/677b718c-e78d-4eb2-8a73-387091b2d982">

It seems their customer support is swamped after launching this watch - hopefully I get my refund…

<img width="169" alt="image" src="https://github.com/user-attachments/assets/819fee17-5220-4515-b02d-755c8c93b606">

**Comment by jk1993** [comment]
I just got my galaxy watch 7 yesterday and have hit the same issue. Managed to get xdrip installed using wear installer but no watch faces and can't even get the complication working. I also haven't been able to get the complication working on my existing watch 4 for at least 6 months or so but that's fine as I use the xdrip watch face but that's no longer possible. I'd say having the BG graph on my watch is probably the main reason for wearing a smartwatch in the first place. I use the sleep tracking etc as well but none of that is important as being able to see what my BG is doing on a graph simply by looking at my wrist.

I did a trade in with my watch 4 so only have a few days before I need to send my old watch back but thinking I might have to cancel and return the new watch if I can't even get my BG displaying on the new watch let alone the graph.

My workaround for my watch 4 if I want to use a different watch face is to use a complication from NightWear which pulls the data from nightscout but this is less than ideal as it can be slow and is dependent on having internet access and my nightscout server being accessible. It also seems unnecessarily hacky when the device uploading the data to nightscout is connected directly to the watch in the first place. So I could probably do this on the new watch too but it feels like a huge downgrade in experience which kind of defeats the point in upgrading the watch.

If anyone is able to help me get at least the complication working on the new watch I'd greatly appreciate it, otherwise I think I'll have to return the new watch.

**Comment by gth001** [comment]
I want to mention, developers don't need a WearOS 5 watch in order to fix this. WearOS 4 will work. Download Watch Face Studio, click add complication, and load it onto your WearOS 4 watch. From there, try to get the xDrip graph loaded into the complication. That's all we need, essentially.

**Comment by rabidsloth** [comment]
I struggled with this yesterday after getting a new watch and found a fix by adding GlucoDataHandler to the mix. It pulls from x-drip and provides complications for watch faces. It's on the play store and open source. Seems to work pretty well, just had to enable 'broadcast locally' and 'broadcast service api' from the inter-app settings in x-drip. This got it working and prevented me from returning the galaxy watch 7.
https://github.com/pachi81/GlucoDataHandler

**Comment by emiltiara** [comment]
I used GlucoDataHandler to show readings on Galaxy Watch 7 - works fine for me, but I have problem with Android Wear Integration on this smartwatch. Does anyone was able to get reading from Wear Collector Service running on Samsung Watch7 and Dexcom G7? If yes could you please share your preferences used for it?

**Comment by malbry** [comment]
I might have a way to enable the xDrip+ watchfaces on a Galaxy Watch 7, even though they are no longer officially supported on those devices. No guarantees though as I don't have a GW7 myself, nor do I use xDrip+. See here: https://youtu.be/EvTGEnPA9PY

**Reply by imshane85** [reply]
> I might have a way to enable the xDrip+ watchfaces on a Galaxy Watch 7, even though they are no longer officially supported on those devices. No guarantees though as I don't have a GW7 myself, nor do I use xDrip+. See here: https://youtu.be/EvTGEnPA9PY

Thanks Malcolm, I can confirm I have this working on a Galaxy Watch 7 running Wear OS5. I followed the instructions exactly as you described - many thanks for this, the historical graphing on the watch face(s) is a very important feature (at least for me!).

![Screenshot_20240829_111539_One UI Watch Home](https://github.com/user-attachments/assets/17c2dd93-17ad-46b1-bce7-6185571da9d8)
![Screenshot_20240829_111901_Settings](https://github.com/user-attachments/assets/589e8ae6-1ed2-4f8f-8c52-d33f8061bf66)
![Screenshot_20240829_111835_Settings](https://github.com/user-attachments/assets/180e3f54-25e5-4eaf-939b-e966e7ff2949)

**Reply by malbry** [reply]
Yay! Thanks for confirming.

**Reply by gth001** [reply]
@imshane85 How have the last 2 weeks been, running xDrip on Watch7? Anything weird happen? Just checking before I trade in my Watch4 and join you.

**Reply by malbry** [reply]
I don't have a GW7 myself, but hopefully someone else can answer here.

**Reply by imshane85** [reply]
> @imshane85 How have the last 2 weeks been, running xDrip on Watch7? Anything weird happen? Just checking before I trade in my Watch4 and join you.

Thankfully smooth sailing, I have not had any issues since installing xDrip+ and the associated watchfaces.

**Reply by nekromantik** [reply]
sadly this does not work on Pixel Watch 3

**Reply by gth001** [reply]
>sadly this does not work on Pixel Watch 3

That's very strange, since both are running WearOS 5. Samsung does customize their build a bit, and Google does release updates faster. Hopefully this doesn't mean the Samsung is going to stop working after the next update.

How exactly did it fail, may I ask? Which step did you get to, and what was the wrong thing it did at that step?

**Reply by imshane85** [reply]
> sadly this does not work on Pixel Watch 3

Damn that's not good. This is precisely why I am not updating my GW7. I'd be curious on what version of wear core services, security software version, and security patch level you are running on the Pixel Watch 3.

**Reply by marcbollinger** [reply]
I couldn't get it to work on the Pixel Watch 3 either :( It fails at the step where (after you install the apk to the watch) in Wear Installer 2, you go to Third Party Apps > Xdrip+ > Install watch faces

**Reply by malbry** [reply]
It turns out that Google have removed the watchface runtime package completely from the Pixel Watch 3. So the method shown in my video won't work on that device :-(

**Reply by imshane85** [reply]
> It turns out that Google have removed the watchface runtime package completely from the Pixel Watch 3. So the method shown in my video won't work on that device :-(

Oh no! Excuse my ignorance, is there a possibility these packages may be removed from the GW7 in future Wear OS versions as the watch is updated? Or is there a method to check packages prior to updating? Thanks!

**Reply by gth001** [reply]
Would it be possible to add this "watchface runtime package" back in? I see someone on XDA forums exported all of the Galaxy Watch 7 .apks and one is named runtime https://xdaforums.com/t/watch-7-ultra-pulled-apps-one-ui.4682929/
<img width="300" alt="image" src="https://github.com/user-attachments/assets/7f2b6664-aa9f-40af-ad36-538ff452c01b"> I don't know much about WearOS though.

**Reply by gth001** [reply]
>is there a possibility these packages may be removed from the GW7 in future Wear OS versions as the watch is updated? Or is there a method to check packages prior to updating? Thanks!

I've disabled updates but I see Samsung has an update available. I'm curious to hear if any brave soul running this workaround has updated. I'm also curious if there is a downgrade method to reflash a watch to original wearos version (I'm guessing no…)

**Reply by Turtlepaw** [reply]
> Would it be possible to add this "watchface runtime package" back in? I see someone on XDA forums exported all of the Galaxy Watch 7 .apks and one is named runtime https://xdaforums.com/t/watch-7-ultra-pulled-apps-one-ui.4682929/ <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/10224533/371946095-7f2b6664-aa9f-40af-ad36-538ff452c01b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzNDM1ODEsIm5iZiI6MTczMTM0MzI4MSwicGF0aCI6Ii8xMDIyNDUzMy8zNzE5NDYwOTUtN2YyYjY2NjQtYWE5Zi00MGFmLWFkMzYtNTM4ZmY0NTJjMDFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDE2NDEyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM3NGI5NGVhODJlNDBmNTcxZGRlNjQ2MWY3N2I4MGY1NDUxM2U0MDU2M2EzMmY3MWNlYmJiNWZhODc1N2MzZDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AvY6NmzfC4ZuBIsK6VA-ft8XVOF9Xhuomd4xee07ZSE"> I don't know much about WearOS though.

Declarative Watch Face Runtime (also referenced as "Watch Face Format Runtime") is used to render [Watch Face Format](https://developer.android.com/training/wearables/wff)

**Comment by gdc27health** [comment]
Hi guys..
I am Graham a 60yo insulin dependent user of xDrip.  Its great on the
phones and tablet but i dont even install it on my watch anymore.
The app to interface is GlucoDataHandler (Michael has done a great job of
providing an overwhelming amount of complications for the wearer.

I have also published a number of faces on the Google Play store.  I am in
the process of updating them to meet the new policies, disclosures and APK
level 33min

Anyway.  That might be a better route.  For get developing the Faces and
Just Pass good complications.

Graham
***@***.***

**Comment by gdc27health** [comment]
I have the new Galaxy watch ultra.   GlucoDataHandler is my go-to for
complications.  Use it on a s22 t send to watch 5 pro, and as follows to
send to watch Ulta from a Flip 6

**Comment by sderaps** [comment]
## DMM Diabetic Watch Faces
Seeing the need for Diabetic Watch Faces with Wear OS 5. I've now completed my project of making Diabetic Watch Faces, that are oversized for the Visually impaired, I also have a Few Diabetic Watch Faces with Graphs.  
*All Watch Faces Require Glucodatahandler App to be installed on phone and watch and work with all wear os watches
*Watch Faces with Graphs require (Blose App installed on Watch (Wear OS 5 only )  and sources (Dexcom Share or Librelinkup only)  

## Sideloading Watch Face Apks or Personalizing .WFS files (Watch Face Studio) Go Here:
https://github.com/sderaps/DMM
## Google Play Store Front for DMM Diabetic Watch Faces Here:
 https://play.google.com/store/apps/dev?id=6551812103351455972&hl=en_US

**Comment by Unrealsweb** [comment]
**WearOS 5 auf der google Smartwatch 2**
My Google Smartwatch 2 received WearOs 5 yesterday. When installing it on the watch, a message appeared saying that the Xdrip software was not compatible with this firmware and might not work. I confirmed this and everything works as usual. I can also change the watch faces of Xdrip+.

**Comment by stereokai** [comment]
What is the status on this issue these days? Thanks

**Reply by Navid200** [reply]
I don't know of any xDrip developer who is working on Wear OS.

**Reply by stereokai** [reply]
Thanks.

**Comment by gth001** [comment]
Status is still the same as the update on the very first line, and not
expected to change. It works great on the Galaxy watch 7 but not the pixel
watch 3.

**Reply by stereokai** [reply]
Thanks.

**Reply by tofof** [reply]
I'm sorry, can you clarify? The first line and your reply here say "it's working great on galaxy watch 7".  However, the video linked to in that solution has a description updated in January that says this technique no longer works on wear os 5 at all, including galaxy watch 7.

**Reply by gth001** [reply]
Wow, that's terrible. Must have been a recent update. News to me. Luckily I still have it installed already. I guess I can never update it again :( That's upsetting. I guess if I lose this watch, I'll have to buy the older model. Or find the old firmware.

**Reply by gth001** [reply]
@malbry - do you know any more details regarding "Update January 2025: this method no longer works on watches that have WearOS 5. This includes the Pixel Watch 3 and the Galaxy Watch 7 and Ultra :-(" ? Was it a particular update? xDrip is still working great on my GW7.

**Reply by stereokai** [reply]
@tofof @gth001 @malbry I can confirm it is still working with Galaxy Watch 7, I just took a leap of faith, bought one and tried it and it works. Also, there are comments on the video after January which report success.

**Reply by imshane85** [reply]
> @malbry - do you know any more details regarding "Update January 2025: this method no longer works on watches that have WearOS 5. This includes the Pixel Watch 3 and the Galaxy Watch 7 and Ultra :-(" ? Was it a particular update? xDrip is still working great on my GW7.

I also have it running on my GW7 per the original video, having said that I am running a very old OS release (refer to details below) and have only updated the watch once since purchase. I would be interested if others have kept the GW7 up to date with patching? @gth001 ?  I am holding off with any further updates as I have some concerns it will break xDrip+ and reverting to a previous OS release is a major pita.

One Ul version
6.0 Watch

Software version
L310XXU1AXK6
SM-L310NZSAKO0

Wear OS version 5.0

Wear core services 1.11.29.603629754

Android security
patch level
1 November 2024

**Reply by gth001** [reply]
I'm on even older firmware than you. I turned off Auto Update and I'm not willing to risk trying it, not unless I'm forced to. 
@stereokai I wonder if since you just got the watch it hasn't updated yet. So that's why you were able to install it.

**Reply by stereokai** [reply]
@imshane85 @gth001  I'm running L310XXU1AXL1 which is the latest update if I'm not mistaken, and that's what came on the watch as I have not performed any update and turned off automatic updates too.

**Comment by malbry** [comment]
I received a report that the method in my video wasn't working on the
latest firmware update for the Galaxy 7 series. But then someone else said
it did work. The problem is that I don't have the Galaxy 7 myself, so I
can't check directly.

**Comment by pkviet** [comment]
Brand new Oneplus watch 3 on wearOs 5 here and unfortunately Wear Installer 2 failed to install the faces ...
I had to resort to using Blose + librelinkup to get the graphs on the watch; it's very unfortunate.
I'm open to run investigations, i'm a c/c++ dev, not an android one but can probably do more than the average user @malbry 
I read on xdrip doc site that unfortunately there is no plan on xdrip+ side to update the wear support to this new wff format.

**Reply by Navid200** [reply]
If anyone can open a PR and show test results that show improvement, I believe they will be considered.
I don't use a watch.  I don't have a WearOS watch.  So, I cannot do anything along those lines myself.
Ideally, a WearOS enthusiast will create a companion app for xDrip and support it.  I believe that will also be considered like WatchDrip was.

**Reply by malbry** [reply]
> Brand new Oneplus watch 3 on wearOs 5 here and unfortunately Wear Installer 2 failed to install the faces ... I had to resort to using Blose + librelinkup to get the graphs on the watch; it's very unfortunate. I'm open to run investigations, i'm a c/c++ dev, not an android one but can probably do more than the average user @malbry I read on xdrip doc site that unfortunately there is no plan on xdrip+ side to update the wear support to this new wff format.

The problem is that Google has completely dropped the old-style watchface runtime in the recent WearOS builds. That affects new Pixel and OnePlus watches, so that watchfaces such as the xDrip+ ones can't be used. There's no way to add the watchface runtime back on those devices. For some reason Samsung has kept the watchface runtime in the latest Series 7 range, albeit not generally accessible except by using Wear Installer 2.  But it's likely that even Samsung will follow Google's lead and remove the old watchface runtime at some point in the future.

Sorry not to have better news.

**Reply by gth001** [reply]
So, I guess we've reached the end of the road for Android watches. I'll keep buying used ones that still work. After that, we can always return to Pebble now that they're going back into production :P https://store.repebble.com/

**Comment by gth001** [comment]
A warning to all: I just let my Galaxy Watch 7 update today and xDrip disappeared :(

"Set Watchface" in Wear Installer 2 doesn't do anything any more :(

**Reply by gth001** [reply]
Anyone know if samsung lets you flash the factory image to get rid of all the updates? I know Pixel Watch 3 lets you via https://developers.google.com/android/images-watch

**Reply by gth001** [reply]
I guess I am OK - I figured out how to use glucodatahandler complications in https://github.com/NightscoutFoundation/xDrip/discussions/3581#discussioncomment-10713357

**Reply by imshane85** [reply]
> A warning to all: I just let my Galaxy Watch 7 update today and xDrip disappeared :(
> 
> "Set Watchface" in Wear Installer 2 doesn't do anything any more :(

I inadvertently followed the same fait - I pressed 'update' on the persistent notification when I was half asleep and couldn't stop it. I knew this day would come but I was not prepared!

The xdrip+ watch face was (is!) way ahead of it's time in my view, especially with the zoomable and fully configurable graphs. I was so heavily reliant on it to manage my bsl. I hope some devs pick it up.

For now I've also got one of the DMM watch faces running with the Bloze (LibreLinkUp) graphs, and whilst it serves a purpose in the interim, its just not the same, and requiring a data connection is a massive pita. I'm still grateful to the devs for building this, but really hanging out for glucodatahandler 2.0 and what it can offer in terms of graphs.

Technology regression sucks.

**Reply by Diabeticliberty** [reply]
I only renewed my old Samsung watch to the Ultra 16 days ago. They will not refund me as they have a 14 day return policy. I wrote them a really terrible review and they bounced straight back at me this morning with the standard 'we sympathise with your issues' and asking me to contact technical support. The problem is the way that their Website is set up you are actively discouraged from pursuing this course of action. I emailed Michael at Glucodatahandler earlier on and explained situations. He said that he also is starting to encounter restrictions which may complicate the application and running of his app. His best suggestion at least for now was to do not install any future watch updates which of course makes every logical sense. Samsung it seems do not want me to be able to use their devices to better manage my own condition but they are very happy to provide me watch faces with cats on or Bitmoji images. FFS why would anyone over the age of 7 want a watch face with cats on it? This of course excludes my ex wife who was obsessed with cats. Suggestions elsewhere on the internet are that Samsung paid Google to force us all into using their really crappy and really buggy watch face generator and watch face store. I GIVE IN I really do. At least for now I have something to be able to refer to which gives me blood glucose data on the hoof. How long this will be good for is up to Mr Samsung and those little darlings at Google.

**Comment by Diabeticliberty** [comment]
I did a Samsung update yesterday on m y Galaxy Watch Ultra. Despite suggesting there would be no major changes, boy oh boy have they changed things. The runtime element that I am told Xdrip+ needs to run on a watch face has gone. Samsung have offered a consolation prize. You can get a watch face with a picture of 2 cats or wait for it..... Bitmoji images WOOOHOOO. Why don't they leave us all the hell alone? In any case I also found Glucodatahandler and managed to install it on phone and watch. Downloaded a couple of watch faces from Diabetic Masked Man in the US and have a couple of watch faces up and running. While these are not representative of the Xdrp+ display which I have come to know and love they do provide all relevant data and I am incredibly grateful to Michael - Glucodatahandler and the American chap Diabetic Masked Man. 

Here is a sneaky peak:

![Glucodatahandler DMM](https://github.com/user-attachments/assets/f22e2db4-26ce-4d55-b53a-db449c34a547)

**Comment by Mustafaorhan0** [comment]
Hello, I bought a watch 7 but the "Sectograph" application that made me buy the watch no longer works in it. I installed it with Wear installer 2 but the watch face did not come. I could only add it as a tab. So what should I do? Can you help me? How can I downgrade? It does not work in Wear OS 5.
![image](https://github.com/user-attachments/assets/d1a240f8-28f1-4732-a19b-7076d82f7331)

**Comment by gth001** [comment]
# FAQ

## I accepted the latest Galaxy Watch 7 update. How do I use Wear Installer 2? How do I use xDrip+ watchface? How do I use non-WFF watchfaces?

You cannot. Samsung has removed the capability. 

## How do I revert the update? 

You cannot. The firmware image isn't anywhere online currently, XDA forums or otherwise. 

## What do I do then? 

### Options

- Buy a compatible watch like Galaxy Watch 6 which still runs WearOS 4
- Keep the Galaxy Watch 7 (or other WearOS 5 watch) and install [GlucoDataHandler - Apps on Google Play](https://play.google.com/store/apps/details?id=de.michelinside.glucodatahandler) to have xDrip+ data on WFF complications. 
  - Optionally, install a watchface specifically designed to show these complications well, like [DMM7 Diabetic Watch Face - Apps on Google Play](https://play.google.com/store/apps/details?id=com.watchfacestudio.DMM7&hl=en_US)
  
## I miss xDrip+ watchface
  
GlucoDataHandler isn't that bad. In some ways it's better. Just try it out.

## I'm here because I googled "wear installer 2" or "watchface runtime" and also have Galaxy Watch 7 issues

Sorry, this page is exclusively about the app xDrip+ so we cannot help with other apps - please ask in your apps github discussions instead TY

## Oh my god why so many emails

I'll close this thread. Samsung did what it did. I don't expect things to change at this point.
